### PR TITLE
chore: move DB access and mutation to preheat and preprocessor DHIS2-12623

### PR DESCRIPTION
**Pull request by teleivo**
## Why

In NTI an event can be imported among other payload combinations given attributeCategoryOptions and **NO** attributeOptionCombo. In this case the NTI needs to figure out what the attributeOptionCombo is. It can do that using the event program category combo and the given attributeCategoryOptions. PreCheckDataRelationsValidationHook previously did just that using the categoryService, thus accessing the DB.

To align with the NTI architecture that logic has been moved to a PreheatSupplier and a PreProcessor.

## What/How

* EventCategoryOptionComboSupplier adds attributeOptionCombo to the preheat for events without a attributeOptionCombo
* EventProgramPreProcessor sets the event.attributeOptionCombo for such events
* PreCheckDataRelationsValidationHook validates the event without accessing the DB

## Other changes

* TrackerImportValidationContext was removed. It held a "cache" used by the PreCheckDataRelationsValidationHook to store attributeOptionCombo ids for events that did not contain one. Hooks like EventCategoryOptValidationHook used the cache to retrieve the attributeOptionCombo for such events. After this PRs changes the TrackerImportValidationContext was just a utility with "getter" style methods that are better placed in the classes owning the returned data i.e. TrackerPreheat, TrackerBundle.
* Added tests (integration, e2e, ...)

### CODE REVIEW

**Review by enricocolasante — APPROVED** [review]
Just a few comments about style

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 1]
```diff
@@ -133,6 +141,122 @@
...
+        return this.cosToCOC.containsKey( categoryOptionComboCacheKey( categoryCombo, categoryOptions ) );
+    }
+
+    /**
+     * Get the category option combo for given category combo and category
+     * options. For the category option combo to exist it has to be stored
+     * before using {@link #putCategoryOptionCombo}.
+     *
+     * @param categoryCombo category combo
+     * @param categoryOptions semi-colon separated list of category options
```
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:207]: It is a semi-colon separated list since that is what we expect in the payload: [event.attributeCategoryOptions](https://github.com/dhis2/dhis2-core/blob/3a6cf1a28322c9c34ac6f8fdaf210615fa4e3567/dhis-2/dhis-services/dhis-service-tracker/src/test/resources/tracker/event_events.json#L12).

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 2]
```diff
@@ -133,6 +141,122 @@
...
+     * @return category option combo identifier
+     */
+    public String getCategoryOptionComboIdentifier( CategoryCombo categoryCombo, String categoryOptions )
+    {
+        CategoryOptionCombo categoryOptionCombo = this.getCategoryOptionCombo(
+            this.cosToCOC.get( categoryOptionComboCacheKey( categoryCombo, categoryOptions ) ) );
+        if ( categoryOptionCombo == null )
```
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:236]: I think its alright, at least here I feel. Client code of this method looks well to me in the end 😋 

https://github.com/dhis2/dhis2-core/pull/9904/files#diff-b50163871f665d8a60e15f8a20ae77161990a73401a61bf1eb410b57f8f5f2a7R126-R128

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/supplier/EventCategoryOptionComboSupplier.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    private boolean hasOnlyExistingCategoryOptions( TrackerPreheat preheat, Set<String> ids )
+    {
+        for ( String id : ids )
+        {
+            if ( preheat.getCategoryOption( id ) == null )
+            {
+                return false;
+            }
+        }
+        return true;
```
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/supplier/EventCategoryOptionComboSupplier.java:149]: This is probably really a matter of style, I have no strong opinion on which one to pick, both work and are easy to understand. What I wonder out of curiosity is, whats the cost associated with creating streams?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 4]
```diff
@@ -133,6 +141,122 @@
...
+
+    /**
+     * Store mapping of category combo + category options identifiers(key) to
+     * category option combo identifiers (value).
+     *
+     * Category options, category option combo identifiers will be in the
+     * idScheme defined by the user on import. Note: different idSchemes for
+     * category combos are not supported.
+     */
+    public void putCategoryOptionCombo( CategoryCombo categoryCombo, Set<CategoryOption> categoryOptions,
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:161]: @enricocolasante I struggled a bit with naming, maybe you have some thoughts.

So we had this map `cosToCOC` before in the validation hook. Now its in the preheat, a bit more robust for when it comes to payload inputs (like ` id1;id1 `, `id2;id1`, `id1;id2`) since I use a set of categoryOption identifiers. So duplicates and order does not matter, also trimming whitespace.

What I wonder is if its confusing that there can be a case in which

`preheat.getCategoryOptionCombo(id)` returns a CategoryOptionCombo
`preheat.getCategoryOptionCombo(categoryCombo, categoryOptions)` returns null

The reason being is that the ClassBasedSupplier will add all CategoryOptionCombo for which there are ids in the payload into the preheat `map`.

The newly added `EventCategoryOptionComboSupplier` will add only CategoryOptionCombo for event payloads that have no CategoryOptionCombo id but CategoryOptions to the preheat `map` and `cosToCOC`. This supplier calls `putCategoryOptionCombo`.

So if

`preheat.getCategoryOptionCombo(categoryCombo, categoryOptions)` returns a CategoryOptionCombo
`preheat.getCategoryOptionCombo(id)` will return the same CategoryOptionCombo

Bottom line, do you think the names are confusing since they are the same, just the signature is different? I added javadocs that explain all of the above. I could not think of a different name. Not sure if that would be necessary.

Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:161]: I think it makes sense to have the same name for overloaded methods.
In the end you are retrieving the same thing just using different parameters.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java [hunk 5]
```diff
@@ -133,6 +141,122 @@
...
+        Set<String> coIds = categoryOptions.stream()
+            .map( getIdentifiers().getCategoryOptionIdScheme()::getIdentifier )
+            .collect( Collectors.toSet() );
+        return Pair.of( categoryCombo.getUid(), coIds );
+    }
+
+    /**
+     * Check if a category option combo for given category combo and category
+     * options has been stored using {@link #putCategoryOptionCombo}. Returns
+     * true if null and a non-null category option combo have been stored.
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:189]: 👍

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:236]: 👍

**Review by enricocolasante — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java:207]: Good point, I will fix that. The payload contains a string field with semi-colon separated identifiers for category options. So this sentence is correct on the other getter where you can pass in the user input (string). Here its just a set of CategoryOption 😄

**Review by stian-sandvold — APPROVED** [review]
